src/analysis/normed_space/hahn_banach/separation.lean

```diff
@@ -0,0 +1,353 @@
+  let g : 𝕜 → E := λ x, a + (x - f a) • (f x₀)⁻¹ • x₀,
+  have hg : continuous g, by continuity,
+  have hcomp : f ∘ g = id, by { ext, simp [hx₀] },
+  have hgfa : g (f a) = a, by simp [hx₀],
+  exact nhds_le_map_nhds hg.continuous_at hcomp hgfa,
+end
+
+variables [normed_group E]
+
+/-- If `A`, `B` are disjoint sets, `A` is compact and `B` is closed then we can surround them while
+keeping them disjoint. -/
+-- TODO: This proof uses the normed group structure of `E`, but it could work for locally convex
+-- topological vector spaces: instead of taking the balls around 0 with radius 1/n, we could show
+-- there must be some convex neighbourhood `W` of 0 which make `A + W` and `B + W` disjoint?
+theorem exists_disjoint_add_ball {A B : set E} (hA : is_compact A) (hB : is_closed B)
+  (disj : disjoint A B) :
+  ∃ ε : ℝ, 0 < ε ∧ disjoint (A + metric.ball 0 ε) (B + metric.ball 0 ε) :=
+begin
+  obtain ⟨u, -, u_pos, u_lim⟩ := exists_seq_strict_anti_tendsto (0 : ℝ),
+  suffices h : ∃ (n : ℕ), disjoint (A + metric.ball 0 (u n)) (B + metric.ball 0 (u n)),
+  { obtain ⟨n, hn⟩ := h,
+    exact ⟨u n, u_pos n, hn⟩ },
+  by_contra' h,
+  simp only [not_disjoint_iff, set.mem_add, metric.mem_ball, dist_zero_right,
+    ← exists_and_distrib_left, ← exists_and_distrib_right, and_assoc] at h,
+  choose z f f' g g' h₁ h₂ h₃ h₄ h₅ h₆ using h,
+  obtain ⟨w, hw, φ, hφ₁, hφ₂ : tendsto (f ∘ _) _ _⟩ := hA.tendsto_subseq h₁,
+  refine disj ⟨hw, mem_of_is_closed_sequential hB (λ n, h₄ (φ n)) _⟩,
+  suffices hfg : tendsto (f - g) at_top (𝓝 0),
+  { simpa only [sub_sub_cancel, sub_zero, comp_app, pi.sub_apply]
+      using hφ₂.sub (hfg.comp hφ₁.tendsto_at_top) },
+  suffices : ∀ n, ∥(f - g) n∥ ≤ 2 * u n,
+  { apply squeeze_zero_norm this,
+    rw ←mul_zero (2 : ℝ),
+    exact u_lim.const_mul (2:ℝ) },
+  intro n,
+  have : f n - g n = g' n - f' n,
+  { rw [sub_eq_iff_eq_add', ←add_sub_assoc, h₆, ←h₃, add_sub_cancel] },
+  rw [pi.sub_apply, this, two_mul],
+  exact (norm_sub_le _ _).trans (add_le_add (h₅ n).le (h₂ n).le),
+end
+
+variables [normed_space ℝ E]
+
+/-- Given a set `C` which is a convex neighbourhood of `0` and a point `x₀` outside of it, there is
+a continuous linear functional `f` separating `x0` and `C`, in the sense that it sends `x₀` to 1 and
+all of `C` to values strictly below `1`. -/
+lemma separate_convex_open_set {C : set E} (zero_mem : (0:E) ∈ C) (hC : convex ℝ C)
+  (hC₂ : is_open C) {x₀ : E} (hx₀ : x₀ ∉ C) :
+  ∃ (f : E →L[ℝ] ℝ), f x₀ = 1 ∧ ∀ x ∈ C, f x < 1 :=
+begin
+  let f : linear_pmap ℝ E ℝ :=
+    linear_pmap.mk_span_singleton x₀ 1 (ne_of_mem_of_not_mem zero_mem hx₀).symm,
+  have hfx₀ : f ⟨(1:ℝ) • x₀, by { dsimp, rw submodule.mem_span_singleton, exact ⟨1, rfl⟩ }⟩ = 1,
+  { simp_rw [linear_pmap.mk_span_singleton_apply, one_smul] },
+  obtain ⟨φ, hφ₁, hφ₂⟩ := exists_extension_of_le_sublinear f (gauge C) _ _ _,
+  { refine ⟨⟨φ, (φ.to_add_monoid_hom.uniform_continuous_of_continuous_at_zero _).continuous⟩, _, _⟩,
```

Comment:
By the way, here's an alternative approach to the middle portion, which uses the standard API for proving continuity of linear maps in normed spaces:
```lean
  have : (C ∩ (-C)) ∈ 𝓝 (0:E) := sorry,
  obtain ⟨r, hr, h⟩ := metric.mem_nhds_iff.1 this,
  obtain ⟨φ, hφ₁, hφ₂⟩ := exists_extension_of_le_sublinear f (gauge C) _ _ _,
  refine ⟨φ.mk_continuous (r⁻¹) _, _, _⟩,
  { intros x,
    rw [real.norm_eq_abs, abs_le, neg_le, ←linear_map.map_neg],
    split; apply (hφ₂ _).trans _,
    { sorry },
    { sorry } },
```